engine/consensus/matching/engine.go

```diff
@@ -797,53 +783,75 @@ func (e *Engine) clearPools(sealedIDs []flow.Identifier) {
 
 	sealed, err := e.state.Sealed().Head()
 	if err != nil {
-		e.log.Error().Err(err).Msg("could not get sealed head")
-		return
+		return fmt.Errorf("could not get sealed head: %w", err)
 	}
 
 	// build a helper function that determines if an entity should be cleared
 	// if it references the block with the given ID
 	missingIDs := make(map[flow.Identifier]bool) // count each missing block only once
-	shouldClear := func(blockID flow.Identifier) bool {
+	shouldClear := func(blockID flow.Identifier) (bool, error) {
 		if e.missing[blockID] >= 1000 {
-			return true // clear if block is missing for 100 seals already
+			return true, nil // clear if block is missing for 100 seals already
 		}
 		header, err := e.headersDB.ByBlockID(blockID)
 		if errors.Is(err, storage.ErrNotFound) {
 			missingIDs[blockID] = true
-			return false // keep if the block is missing, but count times missing
+			return false, nil // keep if the block is missing, but count times missing
 		}
 		if err != nil {
-			e.log.Error().Err(err).Msg("could not check block expiry")
-			return false // keep if we can't check the DB for the referenced block
+			return false, fmt.Errorf("could not check block expiry: %w", err)
 		}
 		if header.Height <= sealed.Height {
-			return true // clear if sealed block is same or higher than referenced block
+			return true, nil // clear if sealed block is same or higher than referenced block
 		}
-		return false
+		return false, nil
+	}
+
+	// The receipts mempool is aware of the Execution Tree structure formed by the execution results.
+	// It supports pruning by height: only results descending from the latest sealed and finalized
+	// result are relevant. Hence, we can prune all results for blocks _below_ the latest block with
+	// a finalized seal. Results of sufficient height for forks that conflict with the finalized fork
+	// are retained in the mempool. However, such orphaned forks do not grow anymore and their
+	// results will be progressively flushed out with increasing sealed-finalized height.
+	err = e.receipts.PruneUpToHeight(sealed.Height)
+	if err != nil {
+		return fmt.Errorf("failed to clean receipts mempool: %w", err)
 	}
 
 	// for each memory pool, clear if the related block is no longer relevant or
 	// if the seal was already built for it (except for seals themselves)
 	for _, result := range e.incorporatedResults.All() {
-		if clear[result.ID()] || shouldClear(result.Result.BlockID) {
+		remove, err := shouldClear(result.Result.BlockID)
+		if err != nil {
+			return fmt.Errorf("failed to evaluate cleaning condition for incorporated results mempool: %w", err)
+		}
+		if clear[result.ID()] || remove {
```

Comment:
no need to do the lookup if we shouldClear.

```suggestion
		if remove || clear[result.ID()] {
```